cmd/virt-launcher/virt-launcher.go

```diff
@@ -347,3 +367,86 @@ func main() {
 
 	log.Log.Info("Exiting...")
 }
+
+// ForkAndMonitor itself to give qemu an extra grace period to properly terminate
+// in case of virt-launcher crashes
+func ForkAndMonitor(qemuProcessCommandPrefix string) error {
+	cmd := exec.Command(os.Args[0], append(os.Args[1:], "--no-fork", "true")...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	done := make(chan struct{}, 2)
+	go func() {
+		io.Copy(os.Stdout, stdout)
+		done <- struct{}{}
+	}()
+
+	go func() {
+		io.Copy(os.Stderr, stderr)
+		done <- struct{}{}
+	}()
+	if err := cmd.Start(); err != nil {
+		log.Log.Reason(err).Error("failed to fork virt-launcher")
+		return err
+	}
+
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+	go func() {
+		for sig := range sigs {
+			log.Log.V(3).Log("signalling virt-launcher to shut down")
+			err := cmd.Process.Signal(syscall.SIGTERM)
+			if err != nil {
+				log.Log.Reason(err).Errorf("received signal %s but can't signal virt-launcher to shut down", sig.String())
+			}
+		}
+	}()
+
+	// stdout and stderr need to be fully read
+	<-done
+	<-done
+
+	// wait for virt-launcher and collect the exit code
+	exitCode := 0
+	if err := cmd.Wait(); err != nil {
+		exitCode = 1
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
+			}
+		}
+		log.Log.Reason(err).Error("dirty virt-launcher shutdown")
+	}
+	// give qemu some time to shut down in case it survived virt-handler
+	pid, _ := virtlauncher.FindPid(qemuProcessCommandPrefix)
+	if pid > 0 {
+		p, err := os.FindProcess(pid)
+		if err != nil {
+			return err
+		}
+		// Signal qemu to shutdown
+		err = p.Signal(syscall.SIGTERM)
+		if err != nil {
+			return err
+		}
+		// Wait for 10 seconds for the qemu process to disappear
+		err = utilwait.PollImmediate(1*time.Second, 10*time.Second, func() (bool, error) {
+			pid, _ := virtlauncher.FindPid(qemuProcessCommandPrefix)
+			if pid == 0 {
+				return true, nil
+			}
+			return false, nil
+		})
+	}
+	// if we saw an exit code > 0 make it visible to the caller
+	if exitCode > 0 {
```

Comment:
would `os.Exit(0)` do the expected thing? any reason not to run `os.Exit(exitCode)` in all cases?